src/SessionManager/ProgramSettings.cc

```diff
@@ -34,6 +38,141 @@ void applyOptionalArgument(T& val, const string& argument_name, const cxxopts::P
+    }
+
+    if (system_settings_json_exists || user_settings_json_exists) {
+        settings.merge_patch(command_line_settings); // force command-line on top of user and sytem
+        SetSettingsFromJSON(settings);
+    }
+}
+
+json ProgramSettings::JSONConfigSettings(const std::string& json_file_path) {
+    std::ifstream ifs(json_file_path, std::ifstream::in);
+    json j;
+    try {
+        j = json::parse(ifs);
+    } catch (json::exception err) {
+        spdlog::warn("Config file {} has problems, please check", json_file_path);
+        spdlog::warn(err.what());
+    }
+    return j;
+}
+
+void ProgramSettings::SetSettingsFromJSON(const json& j) {
+    ValidateJSON(j); // check for invalid keys, store them in std::set for easy counting
+    // logic here is
+    // value = check invalid, if invalid then default value, otherwise check json and assign its value, otherwise default
+    if (!invalid_keys.count("verbosity")) {
+        verbosity = j.value("verbosity", verbosity);
+    }
+    if (!invalid_keys.count("no_log")) {
+        no_log = j.value("no_log", no_log);
+    }
+    if (!invalid_keys.count("log_performance")) {
+        log_performance = j.value("log_performance", log_performance);
+    }
+    if (!invalid_keys.count("log_protocol_messages")) {
+        log_protocol_messages = j.value("log_protocol_messages", log_protocol_messages);
+    }
+    if (!invalid_keys.count("no_http")) {
+        no_http = j.value("no_http", no_http);
+    }
+    if (!invalid_keys.count("no_browser")) {
+        no_browser = j.value("no_browser", no_browser);
+    }
+    if (!invalid_keys.count("host")) {
+        host = j.value("host", host);
+    }
+    if (!invalid_keys.count("port")) {
+        port = j.value("port", port);
+    }
+    if (!invalid_keys.count("grpc_port")) {
+        grpc_port = j.value("grpc_port", grpc_port);
+    }
+    if (!invalid_keys.count("omp_threads")) {
+        omp_thread_count = j.value("omp_threads", omp_thread_count);
+    }
+    if (!invalid_keys.count("top_level_folder")) {
+        top_level_folder = j.value("top_level_folder", top_level_folder);
+    }
+    if (!invalid_keys.count("frontend_folder")) {
+        frontend_folder = j.value("frontend_folder", frontend_folder);
+    }
+    if (!invalid_keys.count("exit_timeout")) {
+        wait_time = j.value("exit_timeout", wait_time);
+    }
+    if (!invalid_keys.count("initial_timeout")) {
+        init_wait_time = j.value("initial_timeout", init_wait_time);
+    }
+    if (!invalid_keys.count("idle_timeout")) {
+        idle_session_wait_time = j.value("idle_timeout", idle_session_wait_time);
+    }
+}
+
+void ProgramSettings::ValidateJSON(const json& j) {
+    for (const auto& key : int_keys) {
+        if (!j.count(key)) {
+            continue;
+        }
+        if (!j[key].is_number_integer()) {
+            spdlog::warn("Config file has problems, please check key with name {}, it's current value is {}", key, j[key]);
```

Comment:
In all these messages "it's" should be "its", and the commas should be full stops.

Reply:
We should also mention the expected type in these messages.

Reply:
Ok.